language/tools/move-cli/src/lib.rs

```diff
@@ -315,22 +299,50 @@ impl OnDiskStateView {
     }
 
     /// Save `module` on disk under the path `module.address()`/`module.name()`
-    fn save_module(&self, module_id: &ModuleId, module_bytes: &[u8]) -> Result<()> {
-        let path = self.get_module_path(module_id);
+    fn save_module(&self, module: &CompiledModule) -> Result<()> {
+        let path = self.get_module_path(&module.self_id());
         if !path.exists() {
             fs::create_dir_all(path.parent().unwrap())?
         }
 
+        let mut module_bytes = vec![];
+        module.serialize(&mut module_bytes)?;
+
         Ok(fs::write(path, &module_bytes)?)
     }
 
-    /// Save all the modules in the local cache
-    /// Returns true if any modules were saved
-    pub fn save_modules(&self) -> Result<bool> {
-        for (id, bytes) in &self.modules {
-            self.save_module(id, &bytes)?
+    /// Save all the modules in the local cache, re-generate mv_interfaces if required.
+    pub fn save_modules<P: Into<PathBuf>>(
+        &self,
+        modules: &[CompiledModule],
+        build_dir_opt: Option<P>,
```

Comment:
It would definitely simplify the code and did come into my mind. The only reason I did not do this in the PR is that I kind of have a mental model of `storage_dir` == `OnDiskStateView` mapping while adding `build_dir` to `OnDiskStateView` also means that the `OnDiskStateView` needs to manage the things produced in the build process (i.e., the mv_interfaces).

Reply:
In fact, there are quite some benefits by merging `build_dir` into `OnDiskStateView`, not only here for the code simplification, but but also for avoid carrying the `args: &Move` (which is a `structopt`) all over the places.

Currently the `args:&Move` is passed to all functions because we need the information of `args.verbose` and `args.build_dir`. Now with commit e198fd7, `OnDiskStateView` knows `build_dir` and `verbose` is passed in as an additional function argument. We no longer need to carry `args:&Move` everywhere.

This makes an important distinction. In Move CLI, only `args:&Move` can `prepare_state()` and to get an `OnDiskStateView`. Given that `args:&Move` disappears before control-flow gets into the command handlers, there is no way re-create another `state` when running the command handlers (e.g., `publish`, `run`, `check`, etc). Hence, reducing the chance of messing up the `storage`.

Reply:
Good points!